src/Elasticsearch.Extensions.Logging/ElasticsearchLogger.cs

```diff
@@ -0,0 +1,359 @@
+
+				_dataProcessor.EnqueueMessage(elasticsearchData);
+			}
+			catch (Exception ex)
+			{
+				Console.Error.WriteLine("ElasticsearchLogger exception: {0}", ex);
+			}
+		}
+
+		private static void AddException(Exception exception, LogEvent logEvent)
+		{
+			// Use the full string of the exception, which includes both the outer stack trace and any
+			// inner exceptions and their stack trace. It also includes any additional values that some
+			// exceptions have in ToString() that is not in Message.
+			var stackTrace = exception.ToString();
+			logEvent.Error = new Error { Type = exception.GetType().FullName, Message = exception.Message, StackTrace = stackTrace };
+		}
+
+		private void AddScopeValues(LogEvent logEvent)
+		{
+			var scopeProvider = ScopeProvider;
+			if (Options.IncludeScopes && scopeProvider != null)
+			{
+				scopeProvider.ForEachScope((scope, innerData) =>
+				{
+					if (logEvent.Labels == null) logEvent.Labels = new Dictionary<string, object>();
+
+					if (logEvent.Scopes == null) logEvent.Scopes = new List<string>();
+
+					var isFormattedLogValues = false;
+					if (scope is IEnumerable<KeyValuePair<string, object>> scopeValues)
+					{
+						foreach (var kvp in scopeValues)
+						{
+							if (kvp.Key == "{OriginalFormat}")
+							{
+								isFormattedLogValues = true;
+								continue;
+							}
+
+							if (CheckTracingValues(logEvent, kvp)) continue;
+
+							logEvent.Labels[kvp.Key] = FormatValue(kvp.Value);
+						}
+					}
+
+					var formattedScope = isFormattedLogValues ? scope.ToString() : FormatValue(scope);
+					logEvent.Scopes.Add(formattedScope);
+				}, logEvent);
+			}
+		}
+
+		private void AddStateValues<TState>(TState state, LogEvent logEvent)
+		{
+			if (state is IEnumerable<KeyValuePair<string, object>> stateValues)
+			{
+				if (stateValues.Count() > 0)
+				{
+					if (logEvent.Labels == null) logEvent.Labels = new Dictionary<string, object>();
+
+					foreach (var kvp in stateValues)
+					{
+						if (kvp.Key == "{OriginalFormat}")
+						{
+							logEvent.MessageTemplate = kvp.Value.ToString();
+							continue;
+						}
+
+						if (CheckTracingValues(logEvent, kvp)) continue;
+
+						logEvent.Labels[kvp.Key] = FormatValue(kvp.Value);
+					}
+				}
+			}
+		}
+
+		private void AddTracing(LogEvent logEvent)
+		{
+			var activity = Activity.Current;
+
+			if (activity != null)
+			{
+				// Unique identifier of the trace.
+				// A trace groups multiple events like transactions that belong together. For example, a user request handled by multiple inter-connected services.
+				logEvent.Trace = new Trace() { Id = activity.RootId };
+
+				// Unique identifier of the transaction.
+				// A transaction is the highest level of work measured within a service, such as a request to a server.
+				var spanId = ExtractW3cSpanIdFromActivityId(activity.Id);
+				logEvent.Transaction = new Transaction() { Id = spanId ?? activity.Id };
```

Comment:
I think setting `Transaction.Id` to `Activity.Current.Id` or to `SpanId` in case it's in W3C format is not correct. As the definition of `transaction` in ecs says (which is in the comment above) "transaction is the highest level of work measured within a service". The thinking behind this is that if you filter based on `transaction.id` you'll see all logs for the given transaction e.g. for a given incoming HTTP request (from the point when it hits the request pipeline until the response is sent back) in a single service. The scope of SpanId or `Activity.Id` is smaller - it'd be something like an outgoing HTTP request or a database call - but each `activity` will have its own id, so if we use those, then the filtering based on `Transaction.Id` will be very different, you'll only see logs for a given span (or activity), but not for the whole transaction.

In [Elastic APM we have the definition of a transaction](https://www.elastic.co/guide/en/apm/get-started/current/transactions.html) which is not the same as a span - `Activity` fits more the definition of the [span](https://www.elastic.co/guide/en/apm/get-started/current/transaction-spans.html). 

We have tracing integration with Serilog and NLog with the Elastic APM Agent (meaning we fill the 2 ecs tracing fields (`Trace.Id` and `Transaction.Id`) based on ids generated by the APM Agent) and in that case we set `Transaction.Id` to the real transaction id - which is the same for all spans within a single service.

Overall I'm afraid it'd be impossible to get a transaction.id just based on `Activity` here and I feel it'd be best to not fill `Transaction.Id` here. We could reiterate on this later, but for now I think it'd be better to just not set it to something which would be technically incorrect. So my suggestion is that we should remove setting the `logEvent.Transaction` part.

Happy to discuss this further, let me know if someone has different opinion.